Add Analysis mode for Sentinel-3

konstntokas opened the pull request:
This PR closes #18.  

It introduces the **analysis mode** for four Sentinel-3 product types:  
- OLCI Level-1 EFR  
- OLCI Level-1 ERR  
- OLCI Level-2 EFR  
- SLSTR Level-2 LST  

Documentation has been updated, and the example notebook `open_sen3.ipynb` now demonstrates opening these products both in native mode and in analysis mode.  

**Notes:**  
- The product type *Sentinel-3 SLSTR Level-1 RBT* requires additional work (see #16).  
- For *Sentinel-3 OLCI Level-2 LRR*, only two sample datasets are available, both of which are broken. We are waiting for the release of valid datasets.

TonioF reviewed (commented):
I did not expect a notebook on sentinel-1 if you don't support it yet.

The plotting of OLCI L1 ERR in native mode looks bad, as it is horizontally strethed. Can you fix that?

Also, I am not a fan of classes that only consist of only capital letters, even if they stand for abbreviations.

Lastly, the open_dataset method in backend.py looks a lot like it should be split into two different methods (maybe even classes), one for opening in analysis and one for native mode. I see though that this was not introduced by this pr.

konstntokas commented on docs/guide.md:
```diff
@@ -33,34 +36,71 @@ form. All bands and quality flags are resampled to a unified, user-provided reso
 Parameters `**kwargs`:
 
 - `resolution`: Target resolution for all spatial data variables / bands.
-  Must be one of `10`, `20`, or `60`. 
```
I removed it here, because `open_dataset(path, engine="eopf-zarr", op_mode="analysis")` is now used for Sentinel-2 and Sentinel-3. For Sentinel-3 the user can set any resolution. For Sentinel-2 it must be on of `10`, `20`, or `60`. I added this in the section Remarks on Specific Sentinel Missions.

konstntokas commented on docs/guide.md:
```diff
@@ -33,34 +36,71 @@ form. All bands and quality flags are resampled to a unified, user-provided reso
+
+#### Sentinel-3
+
+Sentinel-3 products are provided on their **native grid mapping**, where each pixel 
+is defined by a latitude/longitude pair, forming a **2D irregular grid**.  
```
The The SLSTR Level-1B product SL_1_RBT is not supported yet, because there are many subgroups in the zarr product, and I need to talk to  @forman first and decide which subgroups will be added to the analysis mode. 

I opened an issue for this: see #16

konstntokas commented on xarray_eopf/amodes/sentinel3.py:
```diff
@@ -2,9 +2,160 @@
+    def transform_dataset(self, dataset: xr.Dataset, **params) -> xr.Dataset:
+        return self.assign_grid_mapping(dataset)
+
+    def convert_datatree(
```
`convert_datatree` is an abstract method in the class `AnalysisMode` in amode.py. I would not like to change this.

konstntokas commented:
> I did not expect a notebook on sentinel-1 if you don't support it yet.

Sentinel-1 products can be opened with `op_mode="native"`. This example was created in the beginning showing that the Zarr product can be opened. The analysis mode for Sentinel-1 will be added later. 

> The plotting of OLCI L1 ERR in native mode looks bad, as it is horizontally strethed. Can you fix that?

I adjusted the `figsize` to better reflect the actual shape, showing that it is a narrow stripe. 

> Also, I am not a fan of classes that only consist of only capital letters, even if they stand for abbreviations.

I updated the class names to use CamelCase formatting.

> Lastly, the open_dataset method in backend.py looks a lot like it should be split into two different methods (maybe even classes), one for opening in analysis and one for native mode. I see though that this was not introduced by this pr.

Note, we need to implement a single `open_dataset` method, since it is a xarray backend and it is required by xarray. The if/else branches could be put in two different methods, but the assignment needs to be done with `open_dataset`. I do not see a large difference except that it would be slightly better structured. I may change this later. in another PR.

TonioF commented:
> > Lastly, the open_dataset method in backend.py looks a lot like it should be split into two different methods (maybe even classes), one for opening in analysis and one for native mode. I see though that this was not introduced by this pr.
> 
> Note, we need to implement a single `open_dataset` method, since it is a xarray backend and it is required by xarray. The if/else branches could be put in two different methods, but the assignment needs to be done with `open_dataset`. I do not see a large difference except that it would be slightly better structured. I may change this later. in another PR.

I reckoned this would be non-trivial, however, a good structure is always what we should strive for. This shouldn't block the approval of this pr, though.
